### DISCUSSION TITLE
v2.5.0

### DISCUSSION THREAD

**Discussion by epwalsh** [allenai/allennlp#5241] [General]
🆕 AllenNLP `v2.5.0` comes with a few big new features and improvements 🆕

There is a whole new module `allennlp.fairness` that contains implementations of fairness metrics, bias metrics, and bias mitigation tools for your models thanks to @ArjunSubramonian. For a great introduction, check out the corresponding chapter of the guide: https://guide.allennlp.org/fairness.

Another major addition is the `allennlp.confidence_checks.task_checklists` submodule, thanks to @AkshitaB, which provides an automated way to run behavioral tests of your models using the [`checklist`](https://github.com/marcotcr/checklist) library.

`BeamSearch` also has several new important features, including an easy to add arbitrary constraints, thanks to @danieldeutsch.

See below for a comprehensive list of updates 👇

### Added 🎉

- Added `TaskSuite` base class and command line functionality for running [`checklist`](https://github.com/marcotcr/checklist) test suites, along with implementations for `SentimentAnalysisSuite`, `QuestionAnsweringSuite`, and `TextualEntailmentSuite`. These can be found in the `allennlp.confidence_checks.task_checklists` module.
- Added `BiasMitigatorApplicator`, which wraps any Model and mitigates biases by finetuning
on a downstream task.
- Added `allennlp diff` command to compute a diff on model checkpoints, analogous to what `git diff` does on two files.
- Meta data defined by the class `allennlp.common.meta.Meta` is now saved in the serialization directory and archive file
  when training models from the command line. This is also now part of the `Archive` named tuple that's returned from `load_archive()`.
- Added `nn.util.distributed_device()` helper function.
- Added `allennlp.nn.util.load_state_dict` helper function.
- Added a way to avoid downloading and loading pretrained weights in modules that wrap transformers
  such as the `PretrainedTransformerEmbedder` and `PretrainedTransformerMismatchedEmbedder`.
  You can do this by setting the parameter `load_weights` to `False`.
  See [PR #5172](https://github.com/allenai/allennlp/pull/5172) for more details.
- Added `SpanExtractorWithSpanWidthEmbedding`, putting specific span embedding computations into the `_embed_spans` method and leaving the common code in `SpanExtractorWithSpanWidthEmbedding` to unify the arguments, and modified `BidirectionalEndpointSpanExtractor`, `EndpointSpanExtractor` and `SelfAttentiveSpanExtractor` accordingly. Now, `SelfAttentiveSpanExtractor` can also embed span widths.
- Added a `min_steps` parameter to `BeamSearch` to set a minimum length for the predicted sequences.
- Added the `FinalSequenceScorer` abstraction to calculate the final scores of the generated sequences in `BeamSearch`. 
- Added `shuffle` argument to `BucketBatchSampler` which allows for disabling shuffling.
- Added `allennlp.modules.transformer.attention_module` which contains a generalized `AttentionModule`. `SelfAttention` and `T5Attention` both inherit from this.
- Added a `Constraint` abstract class to `BeamSearch`, which allows for incorporating constraints on the predictions found by `BeamSearch`,
  along with a `RepeatedNGramBlockingConstraint` constraint implementation, which allows for preventing repeated n-grams in the output from `BeamSearch`.
- Added `DataCollator` for dynamic operations for each batch.

### Changed ⚠️

- Use `dist_reduce_sum` in distributed metrics.
- Allow Google Cloud Storage paths in `cached_path` ("gs://...").
- Renamed `nn.util.load_state_dict()` to `read_state_dict` to avoid confusion with `torch.nn.Module.load_state_dict()`.
- `TransformerModule.from_pretrained_module` now only accepts a pretrained model ID (e.g. "bert-base-case") instead of
  an actual `torch.nn.Module`. Other parameters to this method have changed as well.
- Print the first batch to the console by default.
- Renamed `sanity_checks` to `confidence_checks` (`sanity_checks` is deprecated and will be removed in AllenNLP 3.0).
- Trainer callbacks can now store and restore state in case a training run gets interrupted.
- VilBERT backbone now rolls and unrolls extra dimensions to handle input with > 3 dimensions.
- `BeamSearch` is now a `Registrable` class.

### Fixed ✅

- When `PretrainedTransformerIndexer` folds long sequences, it no longer loses the information from token type ids.
- Fixed documentation for `GradientDescentTrainer.cuda_device`.
- Re-starting a training run from a checkpoint in the middle of an epoch now works correctly.
- When using the "moving average" weights smoothing feature of the trainer, training checkpoints would also get smoothed, with strange results for resuming a training job. This has been fixed.
- When re-starting an interrupted training job, the trainer will now read out the data loader even for epochs and batches that can be skipped. We do this to try to get any random number generators used by the reader or data loader into the same state as they were the first time the training job ran.
- Fixed the potential for a race condition with `cached_path()` when extracting archives. Although the race condition
  is still possible if used with `force_extract=True`.
- Fixed `wandb` callback to work in distributed training.
- Fixed `tqdm` logging into multiple files with `allennlp-optuna`.

## Commits

b92fd9a7 Contextualized bias mitigation (#5176)
aa52a9a0 Checklist fixes (#5239)
62067973 Fix tqdm logging into multiple files with allennlp-optuna (#5235)
b0aa1d45 Generalize T5 modules (#5166)
5b111d08 tick version for nightly release
39d7e5ae Make BeamSearch Registrable (#5231)
c0142320 Add constraints to beam search (#5216)
98dae7f4 Emergency fix. I forgot to take this out.
c5bff8ba Fixes Checkpointing (#5220)
3d5799d8 Roll backbone (#5229)
babc450d Added `DataCollator` for dynamic operations for each batch. (#5221)
d97ed401 Bump checklist from 0.0.10 to 0.0.11 (#5222)
12155c40 fix race condition when extracting files with cached_path (#5227)
d6629772 cancel redundant GH Actions workflows (#5226)
2d8f3904 Fix W&B callback for distributed training (#5223)
59df2ad3 Update nr-interface requirement from <0.0.4 to <0.0.6 (#5213)
3e1b553b Bump black from 20.8b1 to 21.5b1 (#5195)
d2840cba save meta data with model archives (#5209)
bd941c6f added shuffle disable option in BucketBatchSampler (#5212)
3585c9fe Implementing abstraction to score final sequences in `BeamSearch` (#5208)
79d16af1 Add a `min_steps` parameter to `BeamSearch` (#5207)
cf113d70 Changes and improvements to how we initialize transformer modules from pretrained models (#5200)
cccb35de Rename sanity_checks to confidence_checks (#5201)
db8ff675 Update transformers requirement from <4.6,>=4.1 to >=4.1,<4.7 (#5199)
fd5c9e4c Bias Metrics (#5139)
d9b19b69 Bias Mitigation and Direction Methods (#5130)
74737373 add diff command (#5109)
d85c5c3a Explicitly pass serialization directory and local rank to trainer in train command (#5180)
96c3caf9 fix nltk downloads in install (#5189)
b1b455a2 improve contributing guide / PR template (#5185)
7a260da9 fix cuda_device docs (#5188)
0bf590df Update Makefile (#5183)
3335700c Default print first batch (#5175)
b533733a Refactor span extractors and unify forward. (#5160)
01b232fb Allow google cloud storage locations for cached_path (#5173)
eb2ae30e Update README.md (#5165)
55efa683 fix dataclasses import (#5169)
a463e0e7 Add way of skipping pretrained weights download (#5172)
c71bb460 improve err msg for PolynomialDecay LR scheduler (#5143)
530dae43 Simplify metrics (#5154)
12f5b0f5 Run some slow tests on the self-hosted runner (#5161)
90915800 Fixes token type ids for folded sequences (#5149)
10400e02 Run checklist suites in AllenNLP (#5065)
d11359ed make dist_reduce_sum work for tensors (#5147)
9184fbcb Fixes Backbone / Model MRO inconsistency (#5148)

<hr /><em>This discussion was created from the release <a href='https://github.com/allenai/allennlp/releases/tag/v2.5.0'>v2.5.0</a>.</em>
